Use Thor for built-in tmp task

p8 opened the pull request:
Currently we use both Thor and Rake for `bin/rails` commands. We eventually want to get all the built-ins task promoted to Thor Commands. This migrates the `tmp` task to Thor.

As `tmp` directories are sometimes hidden by editors, I've used `temp` instead.

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

p8 committed:
Use Thor for built-in tmp task

Currently we use both Thor and Rake for `bin/rails` commands.
We eventually want to get all the built-ins task promoted to Thor Commands.
This migrates the `tmp` task to Thor.

Co-authored-by: Juan Vásquez <javasgon@gmail.com>
Co-authored-by: Jonathan Hefner <jonathan@hefner.pro>

jonathanhefner commented on railties/lib/rails/commands/temp/temp_command.rb:
```diff
@@ -0,0 +1,79 @@
+        Rails::Command.invoke "tmp:storage:clear"
+      end
+
+      # desc :create, "Create tmp directories for cache, sockets, and pids"
```
I think we can uncomment this.  It's not commented out for the current Rake task.

p8 replied:
Done.

jonathanhefner commented on railties/lib/rails/commands/temp/temp_command.rb:
```diff
@@ -0,0 +1,79 @@
+      end
+    end
+
+    module Temp
+      class CacheCommand < Base # :nodoc:
```
Unfortunately, putting this module in `railties/lib/rails/commands/temp/temp_command.rb` means [`lookup`](https://github.com/rails/rails/blob/e23e1718ead861ae43004f5706ddad780a440359/railties/lib/rails/command/behavior.rb#L36-L53) won't find these commands (because [`namespaces_to_paths`](https://github.com/rails/rails/blob/e23e1718ead861ae43004f5706ddad780a440359/railties/lib/rails/command/behavior.rb#L70-L80) won't return the right path):

```console
$ bin/rails tmp:screenshots:clear
Unrecognized command "tmp:screenshots:clear" (Rails::Command::UnrecognizedCommandError)
Did you mean?  tmp:sockets:clear

$ bin/rails tmp:sockets:clear
Unrecognized command "tmp:sockets:clear" (Rails::Command::UnrecognizedCommandError)
Did you mean?  tmp:cache:clear

$ bin/rails tmp:cache:clear
Unrecognized command "tmp:cache:clear" (Rails::Command::UnrecognizedCommandError)
Did you mean?  tmp:clear
```

(But "Did you mean?" can subsequently suggest them because [it uses `printing_commands`](https://github.com/rails/rails/blob/e23e1718ead861ae43004f5706ddad780a440359/railties/lib/rails/command.rb#L43).)


So each command needs to go into a dedicated subdirectory.

p8 replied:
I've extracted these to separate files.

jonathanhefner commented on railties/test/commands/temp_test.rb:
```diff
@@ -0,0 +1,47 @@
+
+  test "tmp:clear clear cache, socket, screenshot, and storage files" do
+    Dir.chdir(app_path) do
+      FileUtils.mkdir_p("tmp/cache")
+      FileUtils.touch("tmp/cache/cache_file")
+
+      FileUtils.mkdir_p("tmp/sockets")
+      FileUtils.touch("tmp/sockets/socket_file")
+
+      FileUtils.mkdir_p("tmp/screenshots")
+      FileUtils.touch("tmp/screenshots/fail.png")
+
+      FileUtils.mkdir_p("tmp/storage/6h/np")
+      FileUtils.touch("tmp/storage/6h/np/6hnp81jvgt42pcfqtlpoy8qshfb0")
```
We can use the `app_file` helper for each of these.

p8 replied:
Nice. I've made the change.

jonathanhefner commented on railties/test/commands/temp_test.rb:
```diff
@@ -0,0 +1,47 @@
+
+      rails "tmp:clear"
+
+      assert_not File.exist?("tmp/cache/cache_file")
+      assert_not File.exist?("tmp/sockets/socket_file")
+      assert_not File.exist?("tmp/screenshots/fail.png")
+      assert_not File.exist?("tmp/storage/6h/np/6hnp81jvgt42pcfqtlpoy8qshfb0")
```
It might be worth adding a private `assert_file` helper for these that also incorporates `app_path` calls.  Then we can drop the `Dir.chdir` block.

p8 replied:
I've added an `assert_dir_cleared` that checks the content of the directory.
To make sure we don't delete the directory, this assertion also fails if the directory no longer exists.

jonathanhefner commented on railties/test/commands/temp_test.rb:
```diff
@@ -0,0 +1,47 @@
+
+  test "tmp:create creates tmp folders" do
+    Dir.chdir(app_path) do
+      rails "tmp:create"
+      assert Dir.exist?("tmp/cache")
+      assert Dir.exist?("tmp/sockets")
+      assert Dir.exist?("tmp/pids")
+      assert Dir.exist?("tmp/cache/assets")
```
I think it would be good to call `FileUtils.rm_rf(tmp_dirs)` before `rails "tmp:create"`.  That way we can be sure we are actually testing creation.

p8 replied:
Done.

jonathanhefner reviewed (commented):
> As `tmp` directories are sometimes hidden by editors, I've used `temp` instead.

I don't think changing the code to suit the editor is the right approach.

Which editor are you seeing this problem in?  Would it possible to add a dot file in the command's directory (like a `.keep` file) to force the editor to recognize the directory?

p8 commented:
> Which editor are you seeing this problem in? Would it possible to add a dot file in the command's directory (like a `.keep` file) to force the editor to recognize the directory?

I'm using Neovim with NerdTree.
I thought this was a more common problem, but I'm probably mistaken.
I've updated my vim config and renamed the files 😄

p8 commented on railties/lib/rails/commands/temp/helpers.rb:
```diff
@@ -0,0 +1,27 @@
+    module Temp # :nodoc:
+      module Helpers
+        private
+          def clear_files_in_tmp_dir(dir)
+            require "fileutils"
+            pattern = Rails::Command.application_root.glob("tmp/#{dir}/[^.]*")
+            FileUtils.rm pattern, verbose: false
+          end
+
+          def clear_files_and_dirs_in_tmp_dir(dir)
+            require "fileutils"
+            pattern = Rails::Command.application_root.glob("tmp/#{dir}/[^.]*")
+            FileUtils.rm_rf pattern, verbose: false
+          end
```
> Do you know if there is reason to not combine these two methods?

The original commit doesn't mention any reason.
https://github.com/rails/rails/commit/985cb441103c79651f12a013326aa15f54fa8182

p8 commented on railties/lib/rails/commands/temp/helpers.rb:
```diff
@@ -0,0 +1,27 @@
+          end
+
+          def create_tmp_dir(dir)
+            require "fileutils"
+            FileUtils.mkdir_p Rails::Command.application_root.join("tmp", dir)
```
All these `Pathname` suggestions are really nice. 😄

JuanVqz commented on railties/lib/rails/commands/tmp/tmp_command.rb:
```diff
@@ -0,0 +1,27 @@
+        clear_tmp_dir "storage"
+      end
+
+       desc :create, "Create tmp directories for cache, sockets, and pids"
+      def create
```
```suggestion
      desc :create, "Create tmp directories for cache, sockets, and pids"
      def create
```

p8 replied:
Thanks!
